hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/DataDrivenGroupCacheManager.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2015-2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.hawkular.alerts.engine.impl;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import javax.annotation.PostConstruct;
+import javax.ejb.EJB;
+import javax.ejb.Singleton;
+import javax.ejb.TransactionAttribute;
+import javax.ejb.TransactionAttributeType;
+
+import org.hawkular.alerts.api.model.condition.CompareCondition;
+import org.hawkular.alerts.api.model.condition.Condition;
+import org.hawkular.alerts.api.model.trigger.Trigger;
+import org.hawkular.alerts.api.model.trigger.TriggerType;
+import org.hawkular.alerts.api.services.DefinitionsEvent;
+import org.hawkular.alerts.api.services.DefinitionsListener;
+import org.hawkular.alerts.api.services.DefinitionsService;
+import org.jboss.logging.Logger;
+
+/**
+ * A helper class to keep track of DataDrivenGroup
+ *
+ * @author Jay Shaughnessy
+ * @author Lucas Ponce
+ */
+@Singleton
+@TransactionAttribute(value = TransactionAttributeType.NOT_SUPPORTED)
+public class DataDrivenGroupCacheManager {
+    private final Logger log = Logger.getLogger(DataDrivenGroupCacheManager.class);
+
+    // The sources with member triggers for the dataId.
+    // - null if dataId is not used in a group trigger condition
+    // - EmptySet if no source members are yet defined
+    Map<CacheKey, Set<String>> sourcesMap = new HashMap<>();
+    // The group triggerIds relevant to the dataId, null if none
+    Map<CacheKey, Set<String>> triggersMap = new HashMap<>();
+
+    @EJB
+    DefinitionsService definitions;
+
+    @PostConstruct
+    public void init() {
+        updateCache();
+
+        definitions.registerListener(new DefinitionsListener() {
+            @Override
+            public void onChange(DefinitionsEvent event) {
+                updateCache();
+            }
+        }, DefinitionsEvent.Type.CONDITION_CHANGE);
+    }
+
+    private synchronized void updateCache() {

```

Comment:
Just a note.
As we commented, when we introduce the distributed listener this method can be optimized as well.
It can maintain a cache of the groups model, reducing queries to C*.
This is not to fix anything now, just a comment for future.